File: src/java/com/kuwata_lab/kwartz/Visitor.java
/**
 *  @(#) Visitor.java
 *  @Id  $Id$
 *  @copyright (C)2005 kuwata-lab.com all rights reserverd
 */
package com.kuwata_lab.kwartz;

public class Visitor {
    public final Object visit(Node node) {
        return node.accept(this);
    }

    //
    public Object visitNode(Node expr)             { return null; }
    public Object visitExpression(Expression expr) { return visitNode(expr); }
    public Object visitStatement(Statement stmt)   { return visitNode(stmt); }
    //
    public Object visitBinaryExpression(BinaryExpression expr)               { return visitExpression(expr); }
    public Object visitArithmeticExpression(ArithmeticExpression expr)       { return visitExpression(expr); }
    public Object visitConcatenationExpression(ConcatenationExpression expr) { return visitExpression(expr); }
    public Object visitRelationalExpression(RelationalExpression expr) { return visitExpression(expr); }
    public Object visitAssignmentExpression(AssignmentExpression expr)       { return visitExpression(expr); }
    public Object visitPostfixExpression(PostfixExpression expr)             { return visitExpression(expr); }
    public Object visitPropertyExpression(PropertyExpression expr)           { return visitExpression(expr); }
    public Object visitLogicalAndExpression(LogicalAndExpression expr)       { return visitExpression(expr); }
    public Object visitLogicalOrExpression(LogicalOrExpression expr)         { return visitExpression(expr); }
    public Object visitConditionalExpression(ConditionalExpression expr)     { return visitExpression(expr); }
    public Object visitFunctionExpression(FunctionExpression expr)           { return visitExpression(expr); }
    //
    public Object visitLiteralExpression(LiteralExpression expr)             { return visitExpression(expr); }
    public Object visitStringExpression(StringExpression expr)               { return visitExpression(expr); }
    public Object visitIntegerExpression(IntegerExpression expr)             { return visitExpression(expr); }
    public Object visitFloatExpression(FloatExpression expr)                 { return visitExpression(expr); }
    public Object visitVariableExpression(VariableExpression expr)           { return visitExpression(expr); }
    public Object visitBooleanExpression(BooleanExpression expr)             { return visitExpression(expr); }
    public Object visitNullExpression(NullExpression expr)                   { return visitExpression(expr); }
    //
    public Object visitBlockStatement(BlockStatement stmt)                   { return visitStatement(stmt); }
    public Object visitPrintStatement(PrintStatement stmt)                   { return visitStatement(stmt); }
    public Object visitExpressionStatement(ExpressionStatement stmt)         { return visitStatement(stmt); }
    public Object visitForeachStatement(ForeachStatement stmt)               { return visitStatement(stmt); }
    public Object visitWhileStatement(WhileStatement stmt)                   { return visitStatement(stmt); }
    public Object visitIfStatement(IfStatement stmt)                         { return visitStatement(stmt); }
    public Object visitElementStatement(ElementStatement stmt)               { return visitStatement(stmt); }
    public Object visitExpandStatement(ExpandStatement stmt)                 { return visitStatement(stmt); }
}

File: lib/kwartz/visitor.rb
###
### visitor.rb
###
### $Id$
###

require 'kwartz/exception'

module Kwartz
   
   module Visitor

      ## don't override the method!
      def visit(node, depth=nil)
         return node.accept(self, depth)
      end
      
      ##############################
      
      def visit_expression(expr, depth=0)
         # nothing
      end
      
      def visit_unary_expression(expr, depth=0)
         visit(expr.child, depth)
      end

      def visit_empty_expression(expr, depth=0)
         visit(expr.child, depth)
      end

      def visit_binary_expression(expr, depth=0)
         visit(expr.left, depth+1)
         visit(expr.right, depth+1)
      end

      def visit_funtion_expression(expr, depth=0)
         expr.arguments.each do |expr|
            visit(expr, depth+1)
         end
      end

      def visit_property_expression(expr, depth=0)
         visit(expr.object, depth+1)
         expr.arguments.each do |expr|
            visit(expr, depth+1)
         end if expr.arguments
      end

      def visit_conditional_expression(expr, depth=0)
         visit(expr.condition)
         visit(expr.left)
         visit(expr.right)
      end

      def visit_literal_expression(expr, depth=0)
         return visit_expression(expr, depth)
      end

      def visit_variable_expression(expr, depth=0)
         return visit_literal_expression(expr, depth)
      end

      def visit_numeric_expression(expr, depth=0)
         return visit_literal_expression(expr, depth)
      end

      def visit_string_expression(expr, depth=0)
         return visit_literal_expression(expr, depth)
      end

      def visit_boolean_expression(expr, depth=0)
         return visit_literal_expression(expr, depth)
      end

      def visit_null_expression(expr, depth=0)
         return visit_literal_expression(expr, depth)
      end


      #####################
      
      def visit_statement(stmt, depth=0)
         # nothing
      end
      
      def visit_print_statement(stmt, depth=0)
         stmt.arguments().each do |expr|
            visit(expr, depth)
         end
      end
      
      def visit_expr_statement(stmt, depth=0)
         visit(stmt.expression())
      end
      
      def visit_block_statement(stmt, depth=0)
         block_stmt = stmt
         block_stmt.statements().each do |st|
            visit(st)
         end
      end
      
      def visit_if_statement(stmt, depth=0)
         visit(stmt.condition())
         visit(stmt.then_block())
         visit(stmt.else_stmt()) if stmt.else_stmt()
      end
      
      def visit_foreach_statement(stmt, depth=0)
         visit(stmt.loopvar())
         visit(stmt.list())
         visit(stmt.body())
      end
      
      def visit_while_statement(stmt, depth=0)
         visit(stmt.condition())
         visit(stmt.body())
      end
      
      def visit_macro_statement(stmt, depth=0)
         visit(stmt.body())
      end
      
      def visit_expand_statement(stmt, depth=0)
         # nothing
      end
      
      def visit_rawcode_statement(stmt, depth=0)
         # nothing
      end
      
   end

end

